src/test/app/DepositAuth_test.cpp

```diff
@@ -724,14 +797,539 @@ struct DepositPreauth_test : public beast::unit_test::suite
+                    "01E034";
+                // Alice can't pay with non-existing credentials
+                env(pay(alice, bob, XRP(100)),
+                    credentials::ids({invalidIdx}),
+                    ter(tecBAD_CREDENTIALS));
+            }
+
+            {  // maria can't pay using valid credentials but issued for
+               // different account
+                env(pay(maria, bob, XRP(100)),
+                    credentials::ids({credIdx}),
+                    ter(tecBAD_CREDENTIALS));
+            }
+
+            {
+                // create another valid credential
+                const char credType2[] = "fghij";
+                env(credentials::create(alice, issuer, credType2));
+                env.close();
+                env(credentials::accept(alice, issuer, credType2));
+                env.close();
+                auto const jv =
+                    credentials::ledgerEntry(env, alice, issuer, credType2);
+                std::string const credIdx2 =
+                    jv[jss::result][jss::index].asString();
+
+                // Alice can't pay with invalid set of valid credentials
+                env(pay(alice, bob, XRP(100)),
+                    credentials::ids({credIdx, credIdx2}),
+                    ter(tecNO_PERMISSION));
+            }
+
+            // Error, duplicate credentials
+            env(pay(alice, bob, XRP(100)),
+                credentials::ids({credIdx, credIdx}),
+                ter(temMALFORMED));
+
+            // Alice can pay
+            env(pay(alice, bob, XRP(100)), credentials::ids({credIdx}));
+            env.close();
+        }
+    }
+
+    void
+    testCredentialsCreation()
+    {
+        using namespace jtx;
+
+        const char credType[] = "abcde";
+        Account const issuer{"issuer"};
+        Account const alice{"alice"};
+        Account const bob{"bob"};
+        Account const maria{"maria"};
+
+        {
+            testcase("Creating / deleting with credentials.");
+
+            Env env(*this);
+
+            env.fund(XRP(5000), issuer, alice, bob);
+            env.close();
+
+            {
+                // both included [AuthorizeCredentials UnauthorizeCredentials]
+                auto jv = deposit::authCredentials(bob, {{issuer, credType}});
+                jv[sfUnauthorizeCredentials.jsonName] = Json::arrayValue;
+                env(jv, ter(temMALFORMED));
+            }
+
+            {
+                // both included [Unauthorize, AuthorizeCredentials]
+                auto jv = deposit::authCredentials(bob, {{issuer, credType}});
+                jv[sfUnauthorize.jsonName] = issuer.human();
+                env(jv, ter(temMALFORMED));
+            }
+
+            {
+                // both included [Authorize, AuthorizeCredentials]
+                auto jv = deposit::authCredentials(bob, {{issuer, credType}});
+                jv[sfAuthorize.jsonName] = issuer.human();
+                env(jv, ter(temMALFORMED));
+            }
```

Comment:
Please add tests for `[Unauthorize, UnauthorizeCredentials]` and `[Authorize, UnauthorizeCredentials]`

Reply:
added